### DISCUSSION TITLE
Cancel subscription in a React.useEffect() callback.

### DISCUSSION THREAD

**Discussion by abrman** [pocketbase/pocketbase#1720] [Q&A] [answered]
It feels weird doing async cleanup in react `useEffect()` but it works.
Is this the proper way to handle subscriptions in React on a component level?

```ts
React.useEffect(() => {
    const unsubscribePromise = pb.collection("notes").subscribe(note.id, handleSubscriptionUpdate);
    return () => {
      (async () => {
        (await unsubscribePromise)();
      })();
    };
  },[note.id]);
```

**Comment by ganigeorgiev** [comment, marked as answer]
Yes, the above seems OK to me. I don't know if there is other way to handle async calls in the `useEffect` callback and cleanup functions.

Alternatively, if the above is the only place, where you have a subscription to `note.id`, then you can use the global unsubcribe method, aka.:

```js
// unsubscribe from all notes/node.id subscriptions
pb.collection("notes").unsubscribe(note.id)
```

**Comment by btalanski** [comment]
Tks @ganigeorgiev  for the support on this. I can subscribe and only unsubscribe from a collection when needed using a state flag.  Some extra logic was causing my component to re-render and it was causing the unsubscribe effect to run when there was no actual subscription in place.

This is my final code:

```
useEffect(() => {
    if(!subscribed) {
      console.log('subscribing to shopping_list collection');
      pb.collection('shopping_list').subscribe('*', (event) => { 
        console.log('shopping_list collection refreshing...', event);
        queryClient.invalidateQueries(['shoppingList']);
      });
      setSubscribed(true);
    }
  },[]);

  useEffect(() => {
    return () => {
      if(subscribed){
        console.log('unsubscribing from shopping_list collection');
        pb.collection('shopping_list').unsubscribe();
      }
    }
  },[subscribed]);
```

Feel free to close this ticket.

**Reply by ganigeorgiev** [reply]
Hm, the above check shouldn't be necessary since calling `unsubscribe()` should do nothing if there are no subscriptions.

You can try adding `unsubscribe().catch((err) => {console.log(err, err.originalError)})` to see what is the actual error (if it is from the unsubscribe call) or check the network/console tabs in the browsers's Inspect tool.

**Reply by ganigeorgiev** [reply]
@btalanski If you are getting 404 error during the `unsubscribe` call, you can also check https://github.com/pocketbase/pocketbase/issues/1751#issuecomment-1414307008. If that's the case, then it is safe to ignore it (eg. by providing an empty `catch` handler) if the client-side code cannot be easily adjusted to avoid the eventual unsubscribe requests race conditions.
